Proposed blog post: Improving the usability of C libraries in Swift (#1255)
apple/swift-org-website: _posts/2025-12-15-improving-usability-of-c-libraries-in-swift.md

```swift
├── Package.swift
└── Sources
    └── WebGPU
        ├── include
        │   ├── webgpu.h
        │   └── module.modulemap
        └── WebGPU.c (empty file)
```

If you reference this `WebGPU` target from elsewhere in the package, you can `import WebGPU` to get access to the C APIs.

## Seeing the results

There are a few ways to see what the Swift interface for a C library looks like. 

* The `swift-synthesize-interface` tool in Swift 6.2+ prints the Swift interface to the terminal.
* Xcode's "Swift 5 interface" counterpart to the `webgpu.h` header will show how the header has been mapped into Swift.

I'm going to do it from the command line, using `swift-synthesize-interface`. From the directory containing `webgpu.h` and `module.modulemap`, run:

```swift
xcrun swift-synthesize-interface -I . -module-name WebGPU -target arm64-apple-macos15 -sdk /Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX26.0.sdk
```

> The leading `xcrun` and the `-sdk` argument with the path is only needed because I'm on a Mac; on other platforms, make sure `swift-synthesize-interface` is in your path. The `-target` operation is the triple provided if you run `swiftc -print-target-info`. For me, it looks like this:

Reviewer: ```suggestion
The leading `xcrun` and the `-sdk` argument with the path is only needed on macOS; on other platforms, make sure `swift-synthesize-interface` is in your path. The `-target` operation is the triple provided if you run `swiftc -print-target-info`. It looks like this:
```

  WGPU_FUNCTION_ATTRIBUTE SWIFT_NAME("WGPUBufferImpl.writeBuffer(self:buffer:bufferOffset:data:size:)");
```

There are three things to notice about this `SWIFT_NAME` string:

* It starts with `WGPUBufferImpl.`, which tells Swift to make this function a member inside `WGPUBufferImpl`.
* I've changed the function name to `writeBuffer`, because we no longer need the `wgpuQueue` prefix to distinguish it from other "write buffer" operations on other types.
* The name of the first argument in parentheses is `self`, which indicates that the `self` argument (in Swift) should be passed as that positional argument to the C function. The other arguments are passed in-order.

Note that this also requires `WGPUBuffer(Impl)` to be imported as a `class`, as we did earlier for `WGPUBindGroupImpl`. Once we've done so, we get a much-nicer Swift API:

```swift
extension WGPUQueueImpl {
  /**
   * Produces a @ref DeviceError both content-timeline (`size` alignment) and d
evice-timeline
   * errors defined by the WebGPU specification.
   */
  public func writeBuffer(buffer: WGPUBuffer!, bufferOffset: UInt64, data: UnsafeRawPointer!, size: Int)
}
```

> I've gone and hacked up the header again, but I didn't have to. In `WebGPU.apinotes`, you can put a `SwiftName` attribute on any entity. For `wgpuQueueWriteBuffer`, it would look like this (in the `Functions` section):

Reviewer: ```suggestion
We've hacked up the header again, but didn't have to. In `WebGPU.apinotes`, you can put a `SwiftName` attribute on any entity. For `wgpuQueueWriteBuffer`, it would look like this (in the `Functions` section):
```

```yaml
- Name: wgpuQueueWriteBuffer
  SwiftName: WGPUQueueImpl.writeBuffer(self:buffer:bufferOffset:data:size:)
```

### Importing functions as properties

`WebGPU.h` has a number of `Get` functions that produce information about some aspect of a type. Here are two for the `WGPUQuerySet` type:

```c
WGPU_EXPORT uint32_t wgpuQuerySetGetCount(WGPUQuerySet querySet) WGPU_FUNCTION_ATTRIBUTE;
WGPU_EXPORT WGPUQueryType wgpuQuerySetGetType(WGPUQuerySet querySet) WGPU_FUNCTION_ATTRIBUTE;
```

With the `SWIFT_NAME` tricks above, we can turn these into "get" methods on `WGPUQuerySet`, like this:

```swift
extension WGPUQuerySetImpl {
    public func getCount() -> UInt32
    public func getType() -> WGPUQueryType
}
```

> That's okay, but it's not what you'd do in Swift. Let's go one step further and turn them into read-only computed properties. To do so, we use the `getter:` prefix on the Swift name we define. I'll skip ahead to the YAML form that goes into API notes:

Reviewer: ```suggestion
That's okay, but it's not what you'd do in Swift. Let's go one step further and turn them into read-only computed properties. To do so, use the `getter:` prefix on the Swift name we define. We'll skip ahead to the YAML form that goes into API notes:
```

* The `swift-synthesize-interface` tool in Swift 6.2+ prints the Swift interface to the terminal.

I'm going to do it from the command line, using `swift-synthesize-interface`. From the directory containing `webgpu.h` and `module.modulemap`, run:

```swift
swift-synthesize-interface -I . -module-name WebGPU -target arm64-apple-macos14 -sdk /Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX26.0.sdk
```

The `-sdk` argument with the path is only needed because I'm on a Mac. The `-target` operation is the triple provided if you run `swiftc -print-target-info`. For me, it looks like this:

```json
{
  "compilerVersion": "Apple Swift version 6.2 (swiftlang-6.2.2.15.4 clang-1700.3.15.2)",
  "target": {
    "triple": "arm64-apple-macosx15.0",
    "unversionedTriple": "arm64-apple-macosx",
    "moduleTriple": "arm64-apple-macos",
    "compatibilityLibraries": [ ],
    "librariesRequireRPath": false
  },
  "paths": {
    "runtimeLibraryPaths": [
      "/Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/lib/swift/macosx",
      "/usr/lib/swift"
    ],

>     "runtimeLibraryImportPaths": [

Reviewer 2: Should we `...` out some of the irrelevant parts to make this shorter?

Author: Yes, good idea

```

WebGPU object types are reference counted, and each object type has corresponding `AddRef` and `Release` functions to increment and decrement the reference count, like this:

```c
WGPU_EXPORT void wgpuBindGroupAddRef(WGPUBindGroup bindGroup) WGPU_FUNCTION_ATTRIBUTE;
WGPU_EXPORT void wgpuBindGroupRelease(WGPUBindGroup bindGroup) WGPU_FUNCTION_ATTRIBUTE;
```

Of course, you can use these functions in Swift exactly how you do in C, making sure to balance out calls to `AddRef` and `Release`, but then it would be every bit as unsafe as C.

We can do better with [`SWIFT_SHARED_REFERENCE`](https://www.swift.org/documentation/cxx-interop/#shared-reference-types). It's a macro (defined in the `<swift/bridging>` header) that can turn a reference-counted C type like the above into an automatically reference-counted `class` in Swift. Here's how we would use it in the header:

```swift
typedef struct SWIFT_SHARED_REFERENCE(wgpuBindGroupAddRef, wgpuBindGroupRelease) WGPUBindGroupImpl* WGPUBindGroup WGPU_OBJECT_ATTRIBUTE;
```

Now, `WGPUBindGroup` gets imported like this:

```swift
public class WGPUBindGroupImpl { }
public typealias WGPUBindGroup = WGPUBindGroupImpl
```

> The extra typealias is a little weird, but overall this is a huge improvement: Swift is treating `WGPUBindGroup` as a class, meaning that it automatically manages retains and releases for you! This is both an ergonomic win (less code to write) and a safety win, because it's eliminated the possibility of mismanaging these instances.

Author: Haha, yes. Thank you
